_posts/2025-04-28-sparse-autodiff.md

```diff
@@ -879,44 +880,59 @@ julia> ncolors(prep)
+In fact, `iter_diff` with $k$ iterations gives rise to a banded Jacobian with $k+1$ bands, for which we can easily verify that the optimal coloring uses as many colors as bands, i.e. $c = k+1$.
+For this particular case, the greedy coloring happens to find the optimal solution.
 
 ### Performance benefits
 
-Here we present a benchmark for a slightly larger input, $n = 1000$ and $k = 10$.
-It can be obtained with the following code:
+Here we present a benchmark for the Jacobian of `iter_diff` with varying $n$ and fixed $k$.
+Our goal is to find out when sparse differentiation becomes relevant.
+Benchmark data can be generated with the following code:
 
 ```julia
 using DifferentiationInterfaceTest
-scen = Scenario{:jacobian,:out}(iter_diff, rand(1000); contexts=(Constant(10),))
-data = benchmark_differentiation([dense_backend, sparse_backend], [scen]; benchmark=:full)
+scenarios = [
+    Scenario{:jacobian, :out}(iter_diff, rand(n); contexts=(Constant(k),))
+    for n in round.(Int, 10 .^ (0:0.3:4))
+]
+data = benchmark_differentiation(
+    [dense_backend, sparse_backend],
+    scenarios;
+    benchmark=:full
+)
 ```
 
-In the table below:
+It gives rise to the following performance curves (lower is better):
 
-- the column "sparse" tells us which backend we are using
-- the column "prepared" tells us whether or not preparation is included in the measurements (for dense AD preparation is essentially trivial)
-- the column "time" contains the execution time in seconds
-- the column "bytes" contains the allocated memory in bytes
+{% include figure.html path="assets/img/2025-04-28-sparse-autodiff/demo/benchmark.png" class="img-fluid" %}
+<div class="caption">
+    Figure 22: Performance benefits of sparsity  
+</div>
 
-| **sparse** | **prepared** | **time**  | **bytes** |
-|-----------:|-------------:|----------:|----------:|
-| false      | true         | 2.732e-02 | 1.679e+08 |
-| false      | false        | 2.183e-02 | 1.679e+08 |
-| true       | true         | 1.923e-04 | 1.943e+06 |
-| true       | false        | 2.995e-03 | 1.323e+07 |
+As we can see on Figure 22, there are three main regimes:
 
-<!-- TODO: update benchmarks to new function (re-run Pluto) -->
+1. For very small inputs, we gain nothing by leveraging sparsity.
+2. For medium-sized inputs, sparsity handling is only useful if we can amortize the cost of detection and coloring.
+3. For very large inputs, even the overhead of detection and coloring is worth paying as part of a sparse Jacobian computation.
 
-As shown in the table, even when we include the overhead of pattern detection and coloring, the sparse backend is around $5 \times$ faster than the dense backend.
-The speedup becomes $100 \times$ once we discard this overhead, which can be amortized over several `jacobian` computations.
+Importantly, sparsity can yield an **asymptotic speedup** and not just a constant one.
```

Comment:
Can we find a better term than "asymptotic speedup"? I have trouble wrapping my head around it, since it assumes some "reference performance" that the speed-up is measured w.r.t.

Reply:
yes, the reference performance is that of standard AD